contentprovider.go

```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
 		return nls.locs[n] >= offset
 	})
 
-	end := int(nls.fileSize)
-	if idx < len(nls.locs) {
-		end = int(nls.locs[idx])
+	start, end := nls.lineBounds(idx + 1)
+	return idx + 1, int(start), int(end)
+}
+
+// lineBounds returns the byte offsets of the start and end of the 1-based
+// lineNumber. The end offset is exclusive and will not contain the line-ending
+// newline. If the line number is out of range of the lines in the file, start
+// and end will be clamped to [0,fileSize].
+func (nls newlines) lineBounds(lineNumber int) (start, end uint32) {
+	// newlines[0] is the start of the 2nd line in data.
+	startIdx := lineNumber - 2
+	endIdx := lineNumber - 1
+
+	if startIdx >= 0 && startIdx < len(nls.locs) {
+		start = nls.locs[startIdx] + 1
+	} else if startIdx < 0 {
+		start = 0
+	} else if startIdx >= len(nls.locs) {
+		start = nls.fileSize
 	}
 
-	start := 0
-	if idx > 0 {
-		start = int(nls.locs[idx-1] + 1)
+	if endIdx >= 0 && endIdx < len(nls.locs) {
+		end = nls.locs[endIdx]
+	} else if endIdx < 0 {
+		end = 0
+	} else if endIdx >= len(nls.locs) {
+		end = nls.fileSize
 	}
 
-	return idx + 1, start, end
+	return start, end
 }
 
 // getLines returns a slice of data containing the lines [low, high).
 // low is 1-based and inclusive. high is exclusive.
```

Comment:
is high 1-based too? Could you please clarify it here?

Reply:
Done in https://github.com/sourcegraph/zoekt/pull/366/commits/77be96841336944a903a7aeec58c1abf38c392c0